### ISSUE TITLE
Training got error

### ISSUE THREAD

**Issue by DyncEric** [breizhn/DTLN#28]
Epoch 00095: val_loss did not improve from -16.76465
2021-01-21 00:24:52.466681: W tensorflow/core/kernels/data/generator_dataset_op.cc:107] Error occurred when finalizing GeneratorDataset iterator: Failed precondition: Python interpreter state is not initialized. The process may be terminated.
         [[{{node PyFunc}}]]

**Comment by liziru** [comment]
> Epoch 00095: val_loss did not improve from -16.76465
> 2021-01-21 00:24:52.466681: W tensorflow/core/kernels/data/generator_dataset_op.cc:107] Error occurred when finalizing GeneratorDataset iterator: Failed precondition: Python interpreter state is not initialized. The process may be terminated.
> [[{{node PyFunc}}]]

I found this problem, too on both tf2.3. Now I choose to train on tf2.2 to test whether there is something wrong with my project.  
It's also possible TensorFlow leads to this problem.

**Comment by DyncEric** [comment]
Thanks,  i not used tf2.2, will try

**Comment by breizhn** [comment]
Did it solve the problem?

**Comment by DyncEric** [comment]
> Did it solve the problem?

No, my Graphics card is TX3090, it's not support tf2.2 and below.

**Comment by Tim5Tang** [comment]
@yudashuixiao1 
my be the reason is the code :
early_stopping = EarlyStopping(monitor='val_loss', min_delta=0, 
            patience=10, verbose=0, mode='auto', baseline=None)
when you val_loss did not improve ,large than patience=10 epoch, training will be stopped.

**Comment by StuartIanNaylor** [comment]
I have just been trying with tf 2.7.1 and yeah think its epoch 80 I have forgot but gutted after patiently waiting for so long :)

I have had a google and some point to
```
Just add this before your program i.e after importing dependencies

gpus = tf.config.experimental.list_physical_devices('GPU') 
tf.config.experimental.set_memory_growth(gpus[0], True)
```

I don't know as I just had a gut feeling 
```
# use the GPU with idx 0
os.environ["CUDA_VISIBLE_DEVICES"]='0'
# activate this for some reproducibility
os.environ['TF_DETERMINISTIC_OPS'] = '1'
```
Might be unnecessary,   DETERMINISTIC_OPS just makes the training repeatable but slower doesn't it and not needed apart repeating the process the same and presume takes more memory?
Its an RTX 3050 8Gb so prob understand why gutted when it got to 80 but thought just let CUDA do its thing as never had to set the GPU idx before.

If this fails I will try the Google

**Comment by jeungmin717** [comment]
@StuartIanNaylor 
Hi, I have some issues on training this nice model.

I followed step by step reading through documentation.
training data created by breizhn/DNS-challenge repository
I get validation loss around 49.29
training loss around 0.0014
epoch_lr : decreases from 0.0010 to 1.5259e-08

The validation loss is over too high.
I don't think the data is corrupt, since I am using the same training_data repository you provided.
Can I get some hint or help? How can I get correct validation loss .
only thing I've changed is total_hours 500 -> 40.
I tried to change dropout/ numLayer in DTLN_model.py but it did not improve.

Thanks in advance.

noisyspeech_synthesizer.cfg I used.

sampling_rate: 16000
audioformat: *.wav
audio_length: 30
silence_length: 0.2
total_hours: 40
snr_lower: -5
snr_upper: 25
randomize_snr: True
target_level_lower: -35
target_level_upper: -15
total_snrlevels: 31
clean_activity_threshold: 0.6
noise_activity_threshold: 0.0
fileindex_start: None
fileindex_end: None
is_test_set: False
noise_dir: ./datasets/noise
speech_dir: ./datasets/clean
noisy_destination: ./training_set/noisy
clean_destination: ./training_set/clean
noise_destination: ./training_set/noise
log_dir: ./logs

**Comment by StuartIanNaylor** [comment]
Totally forgot what I got now but probably because you have changed from 500 -> 40.

As when valaidation is much higher than train loss usually its a sign of overfitting and you don't have enough data variance in training.

Your model id working fine on what it has been trained on but external data such as validation is more so likely not enough or varied enough training data

**Comment by jeungmin717** [comment]
@StuartIanNaylor 
Thanks for fast reply,
I've tried both 40h and 500h ( both created automatically by python script provided (noisyspeech_synthesizer_multiprocessing.py)
validation test data was also splitted by script ( split_dns_corpus.py).
I also think this is overfitting too, but I cannot find any solution beside managing data set

**Comment by StuartIanNaylor** [comment]
I tell it what it might be as git clone breizhn not the microsoft one
https://github.com/breizhn/DNS-Challenge and create the dataset as I remember at 1st the dataset I was creating was garbage but as per usual have forgot.
I have MS and apols for the amnesia

**Comment by jeungmin717** [comment]
@StuartIanNaylor 
I am still struggling with val loss in training DTLN
I've checked data set randomly and had no problem.

Can I figure out which one of 1st dataset is garbage?
Or do you get correct val loss, if you run run_training.py? have you changed any configurations?

Sorry for many questions, have been struggling a while

**Comment by StuartIanNaylor** [comment]
All I remember is to use the breizhn fork and the exact same instructtions 1st.

> Training data preparation:
> Clone the forked DNS-Challenge [repository](https://github.com/breizhn/DNS-Challenge). Before cloning the repository make sure git-lfs is installed. Also make sure your disk has enough space. I recommend downloading the data to an SSD for faster dataset creation.
> 
> Run noisyspeech_synthesizer_multiprocessing.py to create the dataset. noisyspeech_synthesizer.cfgwas changed according to my training setup used for the DNS-Challenge.
> 
> Run split_dns_corpus.pyto divide the dataset in training and validation data. The classic 80:20 split is applied. This file was added to the forked repository by me.

Do exactly from the breizhn fork as above and it will work and that will give you a datum to do a custom one.

**Comment by steven8274** [comment]
> 

Maybe you're right.I met the same error just after 10 not improving epochs.

**Comment by steven8274** [comment]
> @yudashuixiao1 Mybe the reason is the code : early_stopping = EarlyStopping(monitor='val_loss', min_delta=0, patience=10, verbose=0, mode='auto', baseline=None) when you val_loss did not improve ,large than patience=10 epoch, training will be stopped.

Maybe you're right.I met the same error just after 10 not improving epochs.
